### PR TITLE
Add support for cross schema joins

### PR DESCRIPTION
**Pull request by Marlinc**
Continuation of #8347 as this is meant for the next feature release.

### CODE REVIEW

**Comment by Marlinc** [comment]
The tests are being flaky and failing due to `Could not authenticate against github.com`. Not sure what I can do about that.

File: tests/TestCase/Database/ConnectionTest.php [hunk 1]
```diff
@@ -962,4 +962,27 @@ public function testSchemaCollection()
...
+     * Tests supportsCrossWith
+     *
+     * @return void
+     */
+    public function testSupportsCrossWith()
+    {
+        $connection = new Connection(ConnectionManager::config('test'));
+        $targetConnection = new Connection(ConnectionManager::config('test'));
+
+        $this->assertFalse($connection->supportsCrossWith($targetConnection), 'The same connection can\'t used in cross');
```
Comment (markstory) [inline tests/TestCase/Database/ConnectionTest.php]: Why not? I there harm in allowing same db 'cross schema' joins?

File: tests/TestCase/Database/ConnectionTest.php [hunk 2]
```diff
@@ -962,4 +962,27 @@ public function testSchemaCollection()
...
+
+        $connection = new Connection(ConnectionManager::config('test'));
+        $targetConnection = new Connection(['name' => 'test2'] + ConnectionManager::config('test'));
+
+        $this->assertTrue($connection->supportsCrossWith($targetConnection), 'Cross should be supported on databases on the same server');
+
+        $connection = new Connection(ConnectionManager::config('test'));
+        $targetConnection = new Connection(['port' => 999999] + ConnectionManager::config('test'));
```
Comment (markstory) [inline tests/TestCase/Database/ConnectionTest.php]: Should the tests also cover different hosts?

File: tests/TestCase/Database/ConnectionTest.php (hunk 1 above)
Comment (Marlinc) [inline tests/TestCase/Database/ConnectionTest.php]: There's no real harm in that, its just that it would cause every 'normal' query to use the new cross schema expression: https://github.com/cakephp/cakephp/pull/8354/files#diff-29389a5b2b355bb26375b37d7b5c3c44R530

File: tests/TestCase/Database/ConnectionTest.php (hunk 2 above)
Comment (Marlinc) [inline tests/TestCase/Database/ConnectionTest.php]: I've added a test for that

File: tests/TestCase/Database/ConnectionTest.php (hunk 1 above)
Comment (markstory) [inline tests/TestCase/Database/ConnectionTest.php]: Would that help simplify some of the other code branches?

Comment (Marlinc) [inline tests/TestCase/Database/ConnectionTest.php]: I'm not sure, where do you think it could help? I presume you're talking about `supportsCrossWith`.

Comment (markstory) [inline tests/TestCase/Database/ConnectionTest.php]: Ok, lets leave it then.

File: tests/TestCase/Database/Expression/CrossSchemaTableExpressionTest.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
```
Comment (lorenzo) [inline tests/TestCase/Database/Expression/CrossSchemaTableExpressionTest.php]: Can you change this to 3.3?

**Comment by Marlinc** [comment]
Should be good to go I think

File: src/Database/Expression/CrossSchemaTableExpression.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        if ($schema instanceof ExpressionInterface) {
+            $schema = $schema->sql($generator);
+        }
+
+        $table = $this->_table;
+        if ($table instanceof ExpressionInterface) {
+            $table = $table->sql($generator);
+        }
+
+        return sprintf('%s.%s', $schema, $table);
```
Comment (chinpei215) [inline src/Database/Expression/CrossSchemaTableExpression.php]: The `quoteIdentifier` setting seems not work. I am not sure that it should be fixed within this method though.
```sql
FROM 
  `users` `Users` 
  LEFT JOIN cakephp2.users `ExternalUsers` ON `Users`.`id` = (`ExternalUsers`.`id`)
```

**Comment by chinpei215** [comment]
But great work!

File: src/Database/Expression/CrossSchemaTableExpression.php (hunk 4 above)
Comment (Marlinc) [inline src/Database/Expression/CrossSchemaTableExpression.php]: I'll look into that, thanks!

**Comment by makallio85** [comment]
Huge thanks for this!

**Comment by lorenzo** [comment]
Thanks @Marlinc :D

File: .travis.yml [hunk 5]
```diff
@@ -49,7 +49,7 @@ before_script:
   - sh -c "if [ '$HHVM' != '1' ]; then phpenv config-rm xdebug.ini; fi"
   - composer self-update
   - if [ -n "$GH_TOKEN" ]; then composer config github-oauth.github.com ${GH_TOKEN}; fi;
-  - composer install --prefer-dist --no-interaction
+  - composer install --prefer-source --no-interaction
```
Comment (chinpei215) [inline .travis.yml]: Why is this changed?

Comment (lorenzo) [inline .travis.yml]: Because travis keeps failing every now and then due to github blocking the download API

**Comment by chinpei215** [comment]
Thank you for teaching me :)
